### ISSUE TITLE
Discrepancy in calculation of {B1, B2, B3, B4} in manuscript

### ISSUE THREAD

**Issue by manupillai308** [Junjue-Wang/FactSeg#1]
In the manuscript, the calculation of **B1, B2, B3** and **B4** is given by 

Bi+1 = Upsamplex2(T(Bi)) + D(F4-i)

Which means for calculating **B2**, we need to upsample B1 then elementwise add the transformed feature map **F3** followed by a **3x3** convolution.

But in the implementation, this is not how **B2** is calculated, nor **B3** and neither **B4**.

The implementation basically upsamples **F4** and then elementwise adds the transformed feature map **F3** on which a **3x3** convolution is performed to obtain **B2**. There is no upsampling of B1.

Is this discrepancy an inadvertant mistake or am I missing something?

PS. According to the original FPN paper, the later is the correct procedure to calculate the feature pyramids.

**Comment by Junjue-Wang** [comment]
> In the manuscript, the calculation of **B1, B2, B3** and **B4** is given by
> 
> Bi+1 = Upsamplex2(T(Bi)) + D(F4-i)
> 
> Which means for calculating **B2**, we need to upsample B1 then elementwise add the transformed feature map **F3** followed by a **3x3** convolution.
> 
> But in the implementation, this is not how **B2** is calculated, nor **B3** and neither **B4**.
> 
> The implementation basically upsamples **F4** and then elementwise adds the transformed feature map **F3** on which a **3x3** convolution is performed to obtain **B2**. There is no upsampling of B1.
> 
> Is this discrepancy an inadvertant mistake or am I missing something?
> 
> PS. According to the original FPN paper, the later is the correct procedure to calculate the feature pyramids.

Thanks for your comment.
The FPN utilizes the former procedure, which you can check in Figure 3 in the original paper.
https://arxiv.org/pdf/1612.03144.pdf

**Comment by manupillai308** [comment]
But according to the implementation of [FPN in SimpleCV](https://github.com/Z-Zheng/SimpleCV/blob/4fa67581441ad150e82b3aa2c394a921f74e4ecd/simplecv/module/fpn.py#L68) module, the forward propagation calculates this:

```python
        ...
        last_inner = getattr(self, self.inner_blocks[-1])(x[-1])
        results = [getattr(self, self.layer_blocks[-1])(last_inner)]
        for feature, inner_block, layer_block in zip(
                x[:-1][::-1], self.inner_blocks[:-1][::-1], self.layer_blocks[:-1][::-1]
        ):
            if not inner_block:
                continue
            inner_top_down = F.interpolate(last_inner, scale_factor=2, mode="nearest")
            inner_lateral = getattr(self, inner_block)(feature)
            last_inner = inner_lateral + inner_top_down
            results.insert(0, getattr(self, layer_block)(last_inner))
       ...
```

Here you can see it basically upsamples the ```last_inner``` which is F4 with a 1x1 convolution and before the 3x3 convolution.
I don't know where exactly am going wrong.

**Comment by Junjue-Wang** [comment]
Yes, the last_inner is successively processed and replaced in: 
```
last_inner = inner_lateral + inner_top_down
```

**Comment by manupillai308** [comment]
So doesn't that mean F4-i is what is upsampled (x2) and then is added with transformed F4-i-1 which is what is then convolved with a 3x3 kernel to create Bi?

Because, in the implementation, ```results``` is the list which holds value of Bi and to calculate the next Bi+1, Bi is not upsampled rather (F4-i + F4-i-1) is?

So, how come in the manuscript, its written Bi-1 is upsampled? I hope SimpleCV's FPN module is what you are using to implement the FactSeg paper 😟

**Comment by Junjue-Wang** [comment]
Yes, I used SimpleCV's FPN module to implement our FactSeg. You can load our pre-trained [weights](https://drive.google.com/file/d/19cCWD3uSZJX_h_carMI1aW6lgAl32qZZ/view) to reproduce the reported accuracy.
